dms/models/res_users.py

```diff
@@ -0,0 +1,91 @@
+
+from typing import Literal  # noqa # pylint: disable=unused-import
+from odoo import api, models
+
+
+class ResUsers(models.Model):
+    # private attributes
+    _inherit = "res.users"
+
+    # default methods
+
+    # fields
+
+    # api.constraints
+
+    # api.model
+
+    # api.depends
+
+    # api.onchange
+
+    # other
+
+    # private
+    @api.model
+    def _get_allowed_directory_ids(self, operation):
+        """
+        Get allowed directory ids for the user based on its inclusion in an access group.
+
+        :param Literal["perm_create", "perm_write", "perm_unlink", "perm_read"] operation: operation to check
+
+        :return: Allowed directory ids
+        :rtype: list[int]
+        """
+        self = self.sudo()
+        all_access = self._has_all_access(field_name=operation)
+
+        # if the user has all access to directories for the current operation, then return all directories
+        if all_access:
+            return [
+                dir["id"] for dir in self.env["dms.directory"].search_read([], ["id"])
+            ]
+
+        my_directory = (
+            self.env["dms.access.group"]
+            .search([("users", "in", self.env.uid), (operation, "=", True)])
+            .directory_ids.filtered(lambda x: not x.is_hidden)
+        )
+        parent_ids = my_directory.parent_id.ids
+
+        # user should be able to see all allowed directories as well as the parent directories of the allowed
+        # directories, and that recursively
+        res_ids = set(my_directory.ids)
+        while parent_ids:
+            res_ids |= set(parent_ids)
+            parent_ids = (
+                self.env["dms.directory"]
+                .search([("id", "in", parent_ids)])
+                .parent_id.ids
+            )
+
+        # When creating a directory, you can't add the directory to the access group, because it doesn't exist yet
+        # In order to allow the user to create the directory anyway, we use a temporary flag
+        just_created_ids = self.env["dms.directory"].search(
+            [("is_first_creation", "=", True)]
+        )
+        res_ids |= set(just_created_ids.ids)
+        just_created_ids.write({"is_first_creation": False})
+
+        return list(res_ids)
+
+    @api.model
+    def _has_all_access(self, user_id=None, field_name="perm_read"):
+        """
+        Check if user has access to all directories for the specified operation.
+
+        :param int user_id: user id to check
+        :param Literal["perm_create", "perm_write", "perm_unlink", "perm_read"] field_name: field name to check
+
+        :return: The access group that has all access
+        :rtype: odoo.model.dms_access_group
+        """
+        return self.env["dms.access.group"].search(
+            [
+                ("users", "in", user_id or self.env.uid),
+                ("has_all_access", "=", True),
+                (field_name, "=", True),
+            ]
+        )
```

Comment:
Although I understand the idea, I do not agree with this change; the initial approach took attachment-type storage into account.